### PR TITLE
Geothermal-reservoir-example

### PR DESCRIPTION
**Pull request by IvarStefansson**
## Proposed changes

Adds an example for a relatively simple subsurface simulation case. In addition to the run script, notable additions are the geometry (two wells and two elliptic fractures) and BCs and ICs based on thermal gradient/hydrostatic pressure/lithostatic stress.

No testing as of yet.

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [ ] The documentation is up-to-date.
- [ ] Static typing is included in the update.
- [ ] This PR does not duplicate existing functionality.
- [ ] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: tests/models/test_fluid_mass_balance.py [hunk 1]
```diff
@@ -946,13 +946,22 @@ def verify_mortar_flux(self, u_known: float):
...
         subdomain, with a small additional jump of aperture at the fracture.
-        The full range is
-        0 (bottom) to -1- aperture (top).
+        The full range is [0 (bottom), -1-aperture (top)].
+
+        Parameters:
+            angle: Angle of gravity with respect to vertical.
+            a: Fracture aperture.
+
+        Asserts:
```
Comment (keileg) [inline tests/models/test_fluid_mass_balance.py:961]: I don't think I have seen this way of making clear what is being tested in our code, but I really like it, at least when it can easily be summarized like this.

File: tests/models/test_fluid_mass_balance.py [hunk 2]
```diff
@@ -946,13 +946,22 @@ def verify_mortar_flux(self, u_known: float):
...
+            That the pressure profile is as expected within a tolerance for both
+            the primary and secondary subdomains.
```
Comment (keileg) [inline tests/models/test_fluid_mass_balance.py]: What is meant by primary and secondary subdomains?

File: tests/models/test_fluid_mass_balance.py (hunk 1 above)
Comment (keileg) [reply tests/models/test_fluid_mass_balance.py:961]: But the question is of course, should we be systematic on when we use this? Cf the next test.

File: tests/models/test_fluid_mass_balance.py [hunk 3]
```diff
@@ -988,13 +997,22 @@ def verify_hydrostatic(self, angle=0, a=1e-1):
...
-        self, discretization, grid_type, num_nodes_mortar, num_nodes_1d
+        self,
+        discretization: Literal["mpfa", "tpfa"],
+        grid_type: Literal["cartesian", "simplex"],
+        num_nodes_mortar: int,
+        num_nodes_1d: int,
     ):
-        """Use homogeneous Neumann boundary conditions on top Dirichlet
-        on bottom.
+        """Use homogeneous Neumann boundary conditions on top Dirichlet on bottom.
```
Comment (keileg) [inline tests/models/test_fluid_mass_balance.py]:
```suggestion
        """Use homogeneous Neumann boundary conditions on top, Dirichlet on bottom.
```

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 4]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+class BoundaryConditionsMechanicsNeumann:
+    """Boundary conditions for the mechanics with Neumann conditions on almost all
+    boundaries.
+
+    The only exception is that internal boundaries are converted to Dirichlet and three
+    points are partly fixed to avoid rigid body motions. We pick maximum z coordinate
+    for all three points. The points have
+            1) mean x and min y coordinate, (fixed in x and z directions)
+            2) mean x and max y coordinate,  (fixed in x and z directions)
+            3) mean y and min x coordinate. (fixed in y and z directions)
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py]: Can we have a short discussion on your experience with this setup in person?

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 5]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+        # and mean x coordinate.
+        point_3 = np.array([x_mean, box["ymin"], z_max])
+        pts = sd.face_centers[:, domain_sides.south]
+        ind_3 = domain_sides.south.nonzero()[0][
+            np.argmin(pp.distances.point_pointset(point_3, pts))
+        ]
+        return [ind_1, ind_2, ind_3]
+
+
+class GravityMagnitude:
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py:398]: I like the idea of this mixin, but I'm not sure about the placement. Should it be somewhere else in the applications - but I don't know where.

Comment (keileg) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:398]: On second thought, I see that it will be used extensively here, so I'm fine with keeping the placement for now.

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 6]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+    """Function returning the domain boundary sides of a given grid."""
+    time_manager: pp.TimeManager
+    """Time manager associated with the model."""
+
+    @property
+    def lithostatic_stress_multipliers(self) -> np.ndarray:
+        """Return multipliers for lithostatic stress.
+
+        Returns:
+            Multipliers for lithostatic stress in the three principal directions.
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py:473]: Ordered as (x, y, z)?

Comment (keileg) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:473]: Does not 'lithostatic' imply that the multiplier in the z-direction is 1?

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 7]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+        Returns:
+            Array of boundary values, with one value for each dimension of the
+                problem, for each face in the subdomain.
+
+        """
+        # Initialize array for stress values.
+        values = np.zeros((3, boundary_grid.num_cells))
+        # Assume zero initial stress state.
+        if self.time_manager.time < 1e-5:
+            return values.ravel("F")
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py:494]: I guess this is motiated by an initialization procedure, but I was really surprise when I saw this. Does this have implications for how this mixin can be used?

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 8]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+        elif material == "solid":
+            gravity = g * self.solid.density
+        else:
+            raise ValueError(f"Unknown region '{material}' for gravity force.")
+        return gravity
+
+
+class LithostaticBoundaryStressValues(GravityMagnitude):
+    """Boundary conditions for the mechanics with lithostatic stress on all
+    boundaries.
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py]: Mention explicitly that the principal stress axes are assumed to coincide with the coordinate axes?

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 9]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+        gravity = self.gravity_force_magnitude("bulk")
+        # Multiply with lithostatic stress multipliers, which can be used to set
+        # different stresses in different directions.
+        gradient = self.lithostatic_stress_multipliers * gravity
+        # Get domain sides and depth at boundary cell centers.
+        domain_sides = self.domain_boundary_sides(boundary_grid)
+        depth = self.depth(boundary_grid.cell_centers)
+
+        # The sign of the stress depends on the side of the domain according to the
+        # direction of the normal vector. Loop over directions.
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py]: Can we make it clearer that this is the normal vector of the domain, not of the individual faces? I was tempted to call it the 'outer normal vector of the domain'.

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py [hunk 10]
```diff
@@ -290,3 +291,366 @@ def bc_values_displacement(self, bg: pp.BoundaryGrid) -> np.ndarray:
...
+        pts = sd.face_centers[:, domain_sides.south]
+        ind_3 = domain_sides.south.nonzero()[0][
+            np.argmin(pp.distances.point_pointset(point_3, pts))
+        ]
+        return [ind_1, ind_2, ind_3]
+
+
+class GravityMagnitude:
+    """Mixin class for gravity magnitude computation for different materials.
+
```
Comment (keileg) [inline src/porepy/applications/boundary_conditions/model_boundary_conditions.py:400]:
```suggestion
The density is taken as the reference density for the material.

```
'Reference density' is not an ideal term, improvements are most welcome.

File: src/porepy/applications/md_grids/model_geometries.py [hunk 11]
```diff
@@ -196,3 +197,273 @@ def create_mdg(self) -> None:
...
+        """Calculate depth of points.
+
+        Parameters:
+            points: Array of points where depth is to be calculated.
+
+        Returns:
+            Depth values for the provided points.
+
+        """
+        return self.domain.bounding_box["zmax"] - points[2, :]
```
Comment (keileg) [inline src/porepy/applications/md_grids/model_geometries.py]: Note to self for discussion: What happens if the gravity does not point in the x3-direction?

File: src/porepy/applications/md_grids/model_geometries.py [hunk 12]
```diff
@@ -196,3 +197,273 @@ def create_mdg(self) -> None:
...
+            "xmin": 0.0,
+            "xmax": x_size,
+            "ymin": 0.0,
+            "ymax": y_size,
+            "zmin": -z_size,
+            "zmax": 0.0,
+        }
+        self._domain = pp.Domain(box)
+
+    def depth(self, points: np.ndarray) -> np.ndarray:
```
Comment (keileg) [inline src/porepy/applications/md_grids/model_geometries.py]: I can see this, together with the client functions for IC and BC, being useful in all kinds of simulations. I suggest we talk through this to see 
i) If this particular function should be made more generally available - does it belong in the main geometry mixin?
ii) To explore how much the code will (may) have to change for models with non-trivial pressure-density relations (not that I will insist on changing this, but I think that an exploration on a blackboard can be worthwhile).

File: src/porepy/applications/md_grids/model_geometries.py [hunk 13]
```diff
@@ -196,3 +197,273 @@ def create_mdg(self) -> None:
...
+    dictionary ``fracture_params`` to the model parameter dictionary. See the property
+    :meth:`fracture_params` for details on the available parameters and their default
+    values.
+
+    If extending to more than two fractures, the user should override all properties
+    defining fracture parameters to return arrays of size (at least) self.num_fractures.
+    The case num_fractures < (size of arrays) is allowed, in which case only the first
+    num_fractures entries are used.
+
+    TODO: Decide whether to replace properties with a defualt dictionary, i.e.,
```
Comment (keileg) [inline src/porepy/applications/md_grids/model_geometries.py]: I lean towards implemetning this. Most of the transparency can be kept by proper documentation in such a unified method.

File: src/porepy/models/solution_strategy.py [hunk 14]
```diff
@@ -523,6 +540,24 @@ def add_nonlinear_diffusive_flux_discretization(
         if discretization not in self._nonlinear_diffusive_flux_discretizations:
             self._nonlinear_diffusive_flux_discretizations.append(discretization)
 
+    def remove_nonlinear_diffusive_flux_discretization(
```
Comment (keileg) [inline src/porepy/models/solution_strategy.py:543]: Check: I suppose you have decided that this will be a sufficiently common operation that it deserves to be in the solution strategy? I trust your judgement here.

**Review by keileg — COMMENTED** [review]
Partial review, I have covered everything except from `geothermal_reservoir.py`, but ran out of time and energy. Hopefully I can complete the review tomorrow.

From what I have seen so far, this will be very useful. There are some topics I think we should discuss in person (see detailed comments), but I have no significant concerns at this stage.

File: src/porepy/examples/geothermal_reservoir.py [hunk 15]
```diff
@@ -0,0 +1,493 @@
...
+logging.basicConfig(level=logging.INFO)
+
+
+class WellBoundaryConditions(pp.PorePyModel):
+    """Class defining boundary conditions for geothermal reservoir models
+    with wells and fractures.
+
+    We impose the well protocols as boundary conditions on the appropriate grids. For
+    the boundaries of the 3d matrix grid (and any fracture grids), we impose hydrostatic
+    pressure and a thermal gradient temperature, corresponding to Dirichlet boundary
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]: Does not this belong elsewhere in the class hierarchy?

File: src/porepy/examples/geothermal_reservoir.py [hunk 16]
```diff
@@ -0,0 +1,493 @@
...
+    conditions for pressure and temperature. We default Dirichlet conditions for all
+    fluxes and all boundaries.
+
+    Super calls in methods `bc_values_pressure` and `bc_values_temperature` suggest that
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:57]: Thanks for being explicit on these matters.

File: src/porepy/examples/geothermal_reservoir.py [hunk 17]
```diff
@@ -0,0 +1,493 @@
...
+
+    if TYPE_CHECKING:
+        well_names: list[str]
+        """List of well tags, e.g. ['injection_well', 'production_well']."""
+
+    def bc_values_pressure(self, bg: pp.BoundaryGrid) -> np.ndarray:
+        """Return boundary values for pressure on all boundaries.
+
+        Parameters:
+            sd: The subdomain for which to return the BC values.
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]:
```suggestion
            bg: The boundary grid for which to return the BC values.
```

File: src/porepy/examples/geothermal_reservoir.py [hunk 18]
```diff
@@ -0,0 +1,493 @@
...
+
+        Returns:
+            The boundary values for pressure on the given subdomain.
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]:
```suggestion
            The boundary values for pressure on the given boundary grid.
```

File: src/porepy/examples/geothermal_reservoir.py [hunk 19]
```diff
@@ -0,0 +1,493 @@
...
+        """
+        sd = bg.parent
+        values = super().bc_values_pressure(bg)  # type: ignore[misc]
+        if self.is_well_grid(sd):
+            well_tag = self.well_names[sd.tags["parent_well_index"]]
+            protocol = self.well_protocols()[well_tag]
+            # Find indices of the well boundary sides.
+            domain_sides = self.domain_boundary_sides(bg)
+            inds = domain_sides.top if self.nd == 3 else domain_sides.north
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:86]:
```suggestion
            # The top of the domain is '.top' in 3d, '.north' in 2d.
            inds = domain_sides.top if self.nd == 3 else domain_sides.north
```
Or something like that - I was puzzled for a little while.

File: src/porepy/examples/geothermal_reservoir.py [hunk 20]
```diff
@@ -0,0 +1,493 @@
...
+                ),
+                "Pa",
+            )
+        return values
+
+    def bc_values_temperature(self, bg: pp.BoundaryGrid) -> np.ndarray:
+        """Return boundary values for temperature on all boundaries.
+
+        Parameters:
+            sd: The subdomain for which to return the BC values.
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]:
```suggestion
            bg: The boundary grid for which to return the BC values.
```

File: src/porepy/examples/geothermal_reservoir.py [hunk 21]
```diff
@@ -0,0 +1,493 @@
...
+
+        Returns:
+            The boundary values for temperature on the given subdomain.
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]:
```suggestion
            The boundary values for temperature on the given boundary grid.
```

File: src/porepy/examples/geothermal_reservoir.py [hunk 22]
```diff
@@ -0,0 +1,493 @@
...
+                self.get_well_value(
+                    protocol["temperatures"],
+                    self.time_manager.schedule,
+                    self.time_manager.time,
+                ),
+                "K",
+            )
+        return values
+
+    def is_well_grid(self, sd: pp.Grid) -> bool:
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:127]: Not for this PR, but should we provide a better way of identifying well subdomains?

File: src/porepy/examples/geothermal_reservoir.py [hunk 23]
```diff
@@ -0,0 +1,493 @@
...
+        current_time: float,
+    ) -> float:
+        """Return the well value at the current time.
+
+        Parameters:
+            values: Array with time-dependent well values.
+            times: Array with time points corresponding to the well values.
+                Linear interpolation is used to find the value at the current time.
+            current_time: Current simulation time.
+
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:151]: Raises..

File: src/porepy/examples/geothermal_reservoir.py [hunk 24]
```diff
@@ -0,0 +1,493 @@
...
+        else:
+            return float(np.interp(current_time, times, values))
+
+    def well_protocols(self) -> dict[str, dict[str, NDArray[np.float64]]]:
+        """Dictionary mapping well tags to well protocols.
+
+        Returns:
+            Dictionary with well protocols, each containing a dictionary with
+            time-dependent temperatures and pressures, with each value being an array of
+            size equal to the number of time steps.
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]: Scheduled time steps? By default, time steps are the numerical time steps,  I would say.

File: src/porepy/examples/geothermal_reservoir.py [hunk 25]
```diff
@@ -0,0 +1,493 @@
...
+            return pp.BoundaryCondition(sd, inds, "neu")
+        else:
+            return super().bc_type_darcy_flux(sd)  # type: ignore[misc]
+
+    def bc_type_fourier_flux(self, sd: pp.Grid) -> pp.BoundaryCondition:
+        """Return Neumann BC for Fourier flux on well grids during first time interval.
+
+        Parameters:
+            sd: The subdomain for which to return the BC type.
+        Returns:
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:252]:
```suggestion

        Returns:
```

File: src/porepy/examples/geothermal_reservoir.py [hunk 26]
```diff
@@ -0,0 +1,493 @@
...
+        is very cheap for 1d well grids.
+
+        """
+        # Determine if we are at the start of the second time interval.
+        t_prev = self.time_manager.time - self.time_manager.dt
+        rediscretize_flow = np.isclose(t_prev, self.time_manager.schedule[1])
+        interval_ind = np.searchsorted(
+            self.time_manager.schedule, self.time_manager.time, side="right"
+        )
+        logger.info(
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]: Granted this is an example, thus it does not set precedence, but is this the right method to do this kind of logging?

File: src/porepy/examples/geothermal_reservoir.py [hunk 27]
```diff
@@ -0,0 +1,493 @@
...
+            inds = domain_sides.top if self.nd == 3 else domain_sides.north
+            return pp.BoundaryCondition(sd, inds, "neu")
+        else:
+            return super().bc_type_fourier_flux(sd)  # type: ignore[misc]
+
+    def before_nonlinear_loop(self) -> None:
+        """Hook called before the nonlinear solver loop.
+
+        Rediscretize flow problem at the onset of second time interval, when injection
+        starts. The tailoring is done to recompute the Darcy flux discretization only at
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py]: I don't see the motivation for this operation; the time saved will be miniscule, right?

File: src/porepy/examples/geothermal_reservoir.py [hunk 28]
```diff
@@ -0,0 +1,493 @@
...
+    injection_pressures = injection_pressures[:schedule_length]
+
+    # Adjust solid values, while using default values for water.
+    solid_values = cast(dict[str, float], pp.solid_values.basalt)
+    solid_values.update(
+        {
+            "dilation_angle": 0.1,  # [rad]
+            # Uncomment next two lines to include elastic fracture deformation, aka
+            # "Barton-Bandis" model for normal fracture deformation.
+            # "fracture_normal_stiffness": 1.1e8,  # [Pa m^-1]
```
Comment (keileg) [inline src/porepy/examples/geothermal_reservoir.py:438]: For curiosity, does the Barton-Bandis model impact non-linear convergence?

**Review by keileg — COMMENTED** [review]
Review of the run script. Nice work!

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py (hunk 6 above)
Comment (IvarStefansson) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:473]: Possibly. I'm fine with enforcing that, but I don't see the point of such an restriction. As is, the class can be used with y as the vertical axis or for other weird user preferences.

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py (hunk 7 above)
Comment (IvarStefansson) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:494]: Not how it can be used, but rather how it should be used. This is intended to be used with zero initial displacements. And yes, that should be at least acknowledged. Would a class level comment be the best solution?

File: src/porepy/applications/md_grids/model_geometries.py (hunk 12 above)
Comment (IvarStefansson) [reply src/porepy/applications/md_grids/model_geometries.py]: Not sure what you mean by ii, but yes, let's discuss.

File: src/porepy/applications/md_grids/model_geometries.py (hunk 13 above)
Comment (IvarStefansson) [reply src/porepy/applications/md_grids/model_geometries.py]: Partial implementation. Let me know if you want me to move the centers as well. If so, should I keep the scaling with domain sizes?

File: src/porepy/examples/geothermal_reservoir.py (hunk 22 above)
Comment (IvarStefansson) [reply src/porepy/examples/geothermal_reservoir.py:127]: Yes.

File: src/porepy/examples/geothermal_reservoir.py (hunk 27 above)
Comment (IvarStefansson) [reply src/porepy/examples/geothermal_reservoir.py]: It feels like a more "proper" way of doing the rediscretization. Let's discuss.

File: src/porepy/examples/geothermal_reservoir.py (hunk 28 above)
Comment (IvarStefansson) [reply src/porepy/examples/geothermal_reservoir.py:438]: Don't know.

File: src/porepy/models/solution_strategy.py (hunk 14 above)
Comment (IvarStefansson) [reply src/porepy/models/solution_strategy.py:543]: It depends on what you think of my example usage. But I think the symmetry with "add_" is a partial justification for "remove_"

File: tests/models/test_fluid_mass_balance.py (hunk 1 above)
Comment (IvarStefansson) [reply tests/models/test_fluid_mass_balance.py:961]: Yes, I think we should add it as "best practice". I don't think it's worth a full coverage effort.

File: src/porepy/examples/geothermal_reservoir.py (hunk 15 above)
Comment (IvarStefansson) [reply src/porepy/examples/geothermal_reservoir.py]: See updates.

File: src/porepy/applications/md_grids/model_geometries.py (hunk 12 above)
Comment (keileg) [reply src/porepy/applications/md_grids/model_geometries.py]: Let's move it. 

also, point ii) can be ignored for this PR.

File: src/porepy/examples/geothermal_reservoir.py (hunk 27 above)
Comment (keileg) [reply src/porepy/examples/geothermal_reservoir.py]: Let's make clearer that the motivation is related to the switch of boundary condition type.

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py (hunk 6 above)
Comment (keileg) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:473]: Good point.

File: tests/models/test_fluid_mass_balance.py (hunk 1 above)
Comment (keileg) [reply tests/models/test_fluid_mass_balance.py:961]: Agreed. Resolve here, document somewhere?

File: src/porepy/examples/geothermal_reservoir.py (hunk 22 above)
Comment (keileg) [reply src/porepy/examples/geothermal_reservoir.py:127]: https://github.com/pmgbergen/porepy/issues/1552

File: src/porepy/applications/boundary_conditions/model_boundary_conditions.py (hunk 7 above)
Comment (keileg) [reply src/porepy/applications/boundary_conditions/model_boundary_conditions.py:494]: Yes, that should do.

File: src/porepy/applications/md_grids/model_geometries.py (hunk 13 above)
Comment (keileg) [reply src/porepy/applications/md_grids/model_geometries.py]: Much better. I'm not sure about the centers, but we should treat them and the minor/major axis equally, I think, since these are the quantities that scale with domain size.

If forced, I will vote for keeping centers and axes as separate functions and keep the scaling. Keeping the scaling is more important to me than to have separate functions. None of these opinions are very strong. 

Let me know if you want to discuss, but if you are ready to make a decision, please go ahead.

**Review by keileg — APPROVED** [review]
Feel free to merge when comments are resolved and let me know if you want further input from my side.
